pallets/fusion/src/lib.rs

```diff
@@ -0,0 +1,1989 @@
+		#[pallet::call_index(7)]
+		#[pallet::weight(T::WeightInfo::create_fusion_currency())]
+		pub fn set_fusion_pool(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			apy: Option<Perbill>,
+			state: Option<FusionPoolState>,
+			nominator: Option<Option<T::AccountId>>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			FusionPools::<T>::try_mutate(pool_id, |maybe_pool| {
+				let pool = maybe_pool.as_mut().ok_or(Error::<T>::PoolNotFound)?;
+
+				ensure!(
+					pool.state != FusionPoolState::Destroying,
+					Error::<T>::PoolIsDestroying
+				);
+
+				if let Some(apy_value) = apy {
+					pool.apy = apy_value;
+				}
+
+				if let Some(state) = state {
+					ensure!(
+						state != FusionPoolState::Destroying,
+						Error::<T>::CannotSetPoolToDestroying
+					);
+					if state == FusionPoolState::Open || state == FusionPoolState::Blocked {
+						ensure!(pool.targets.len() > 0, Error::<T>::PoolIsNotNominating);
+						let currency = FusionCurrencies::<T>::get(pool.currency_id)
+							.ok_or(Error::<T>::CurrencyNotFound)?;
+						ensure!(!currency.is_destroyed, Error::<T>::CurrencyDestroyed);
+					}
+
+					pool.state = state;
+				}
+
+				if let Some(nominator) = nominator.clone() {
+					pool.nominator = nominator;
+				}
+
+				// Emit an event for pool update
+				Self::deposit_event(Event::PoolSet {
+					pool_id,
+					apy,
+					state,
+					nominator,
+				});
+
+				Ok(())
+			})
+		}
+
+		/// Deletes a pool
+		/// Called once to set the pool to destroying
+		/// Called a second time when everything is cleaned to actually destroy it
+		#[pallet::call_index(8)]
+		#[pallet::weight(T::WeightInfo::create_fusion_currency())]
+		pub fn delete_fusion_pool(origin: OriginFor<T>, pool_id: PoolId) -> DispatchResult {
+			ensure_root(origin)?;
+
+			FusionPools::<T>::try_mutate(pool_id, |maybe_pool| -> DispatchResult {
+				let pool = maybe_pool.as_mut().ok_or(Error::<T>::PoolNotFound)?;
+
+				if pool.state != FusionPoolState::Destroying {
+					pool.state = FusionPoolState::Destroying;
+					Self::deposit_event(Event::PoolDestroying { pool_id });
+				} else {
+					Self::check_and_cleanup_pool(&pool)?;
+				}
+
+				Ok(())
+			})
+		}
+
+		/// Nominates a list of validators for a given pool.
+		#[pallet::call_index(9)]
+		#[pallet::weight(T::WeightInfo::create_fusion_currency())]
+		pub fn nominate(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			targets: BoundedVec<T::AccountId, T::MaxTargets>,
+		) -> DispatchResult {
+			// Check if the origin is root, if not, check if it's a signed origin.
+			let is_root = ensure_root(origin.clone()).is_ok();
+			let who = if is_root {
+				None
+			} else {
+				Some(ensure_signed(origin)?)
+			};
+
+			// Fetch the pool and ensure it exists
+			FusionPools::<T>::try_mutate(pool_id, |pool_opt| -> DispatchResult {
+				let pool = pool_opt.as_mut().ok_or(Error::<T>::PoolNotFound)?;
+
+				// If the caller is not root, ensure it's the nominator of the pool
+				if let Some(caller) = who {
```

Comment:
better naming is needed.

Reply:
What do you mean ?

Reply:
It's a bit unclear what's going on here. 
```rust
let is_root = ensure_root(origin.clone()).is_ok();
let who = if is_root {
   None
} else {
    Some(ensure_signed(origin)?)
};

if let Some(caller) = who {
    ensure!(Some(&caller) == pool.nominator.as_ref(), Error::<T>::NotAuthorized);
}			
```
I will come up with a better option what's I am done reviewing

Reply:
root can do whatever, if it's a regular address, only nominator can change nominations